magma/smart/smart_bits.py

```diff
@@ -0,0 +1,526 @@
+            return f"{op_name}Reduce"
+
+    def __init__(self, op, operand):
+        reduction = _SmartReductionOpExpr._ReductionOp(op)
+        super().__init__(reduction, operand)
+
+    def resolve(self, context):
+        context = Context(None, self)
+        self._resolve_args(context)
+        self._width_ = 1
+        self._signed_ = all(arg._signed_ for arg in self._args)
+
+    def eval(self):
+        args = self._eval_args()
+        fn = sint if self._signed_ else uint
+        return fn(self.op(*args))
+
+
+class _SmartNAryContextualOpExr(_SmartOpExpr):
+    def __init__(self, op, *args):
+        super().__init__(op, *args)
+
+    def resolve(self, context):
+        self._resolve_args(context)
+        to_width = context.max_width()
+        args = (_extend_if_needed(arg, to_width) for arg in self._args)
+        self._update(*args)
+        self._width_ = to_width
+        self._signed_ = all(arg._signed_ for arg in self._args)
+
+    def eval(self):
+        args = self._eval_args()
+        if self._signed_:
+            args = (sint(arg) for arg in args)
+        else:
+            args = (uint(arg) for arg in args)
+        return self.op(*args)
+
+
+class _SmartBinaryOpExpr(_SmartNAryContextualOpExr):
+    def __init__(self, op, loperand, roperand):
+        super().__init__(op, loperand, roperand)
+
+
+class _SmartUnaryOpExpr(_SmartNAryContextualOpExr):
+    def __init__(self, op, operand):
+        super().__init__(op, operand)
+
+
+class _SmartComparisonOpExpr(_SmartOpExpr):
+    def __init__(self, op, loperand, roperand):
+        super().__init__(op, loperand, roperand)
+
+    def resolve(self, context):
+        context = Context(None, self)
+        self._resolve_args(context)
+        to_width = max(arg._width_ for arg in self._args)
+        args = (_extend_if_needed(arg, to_width) for arg in self._args)
+        self._update(*args)
+        self._width_ = 1
+        self._signed_ = False
+
+    def eval(self):
+        args = self._eval_args()
+        signed_args = all(arg._signed_ for arg in self._args)
+        fn = sint if signed_args else uint
+        args = (fn(arg) for arg in args)
+        return uint(self.op(*args))
+
+
+class _SmartShiftOpExpr(_SmartOpExpr):
+    def __init__(self, op, loperand, roperand):
+        super().__init__(op, loperand, roperand)
+
+    def resolve(self, context):
+        loperand, roperand = self._args
+        loperand.resolve(context)
+        self_context = Context(None, self)
+        roperand.resolve(self_context)
+        to_width = max(arg._width_ for arg in self._args)
+        args = (_extend_if_needed(arg, to_width) for arg in self._args)
+        self._update(*args)
+        self._width_ = to_width
+        self._signed_ = all(arg._signed_ for arg in self._args)
+
+    def eval(self):
+        args = self._eval_args()
+        if self._signed_:
+            args = (sint(arg) for arg in args)
+        else:
+            args = (uint(arg) for arg in args)
```

Comment:
The mapping of `_signed_` to `uint` or `sint` can probably be wrapped up in a helper function since it's the same for all these `eval` methods.

Reply:
Also it looks like `_SmartReductionOpExpr`, `_SmartNAryContextualOpExr`, and `_SmartShiftOpExpr` all have the same `eval` function that could be reused.  

`_SmartComparisonOpExpr` looks similar, but `signed_args` is used instead of `_signed_` (so could share some helper function like `_eval(self._signed_)` and `_eval(signed_args)`

Reply:
Agreed. In general the `eval` and `resolve` methods could be cleaned up and unified I think. However, given that it is a little bit fragile, I want to make sure the logic is right, and then I will think about a unification of these.